frontend/app/actions/RegistrationUri.scala

```diff
@@ -9,43 +9,38 @@ object RegistrationUri {
   val REFERRER_HEADER: String = "referer"
   val CAMPAIGN_SOURCE: String = "MEM"
 
+  val referingPageCodes = Map(
+    controllers.routes.Info.supporter().toString() -> "SUPUK",
+    controllers.routes.Info.supporterUSA().toString() -> "SUPUS",
+    controllers.routes.Info.offersAndCompetitions().toString()-> "COMP",
+    controllers.routes.WhatsOn.list().toString() -> "EVT",
+    "/" -> "HOME"
+  )
+
+  val campaignTierCodes = Map[String, String](
+    controllers.routes.Joiner.joinPaid(Tier.Supporter).toString() -> "SUP",
+    controllers.routes.Joiner.joinPaid(Tier.Partner).toString() -> "PAR",
+    controllers.routes.Joiner.joinPaid(Tier.Patron).toString() -> "PAT",
+    controllers.routes.Joiner.joinPaid(Tier.Friend).toString() -> "FRI"
+  )
+
   def parse(request: RequestHeader) = {
     val redirectUrl: String = Config.idWebAppRegisterUrl(request.uri)
-    extractCampaignCode(request.headers(REFERRER_HEADER), request.path) match {
-      case Some(campaignCode) => redirectUrl.concat(s"&INTCMP=${campaignCode.get}")
-      case _ => redirectUrl
-    }
+    val campaignCode = extractCampaignCode(request.headers.get(REFERRER_HEADER), request.path)
+    redirectUrl.concat(s"&INTCMP=$campaignCode")
   }
 
-  private def extractCampaignCode(referer: String, path: String) = {
+  private def extractCampaignCode(refererOpt: Option[String], path: String) : String = {
+
+    val referingCode = refererOpt.flatMap(getReferingPageCode)
+    val campaignTier: Option[String] = campaignTierCodes.get(path)
+
+    (Seq(CAMPAIGN_SOURCE) ++ referingCode ++ campaignTier).mkString("_")
```

Comment:
Does it rely on implicit conversion from `Option` to `Seq`?

If so, then we can easily have an inconsistent code without a referingCode.

What about a simple `Option`-based for-comprehension instead?